web/src/buddy_list.js

```diff
@@ -55,16 +87,259 @@ class BuddyListConf {
+            hide_headers,
+            subscriber_count,
+            other_users_count,
+            has_inactive_users_matching_view,
+            has_inactive_other_users,
+            total_user_count,
+        };
+    }
 
     populate(opts) {
         this.render_count = 0;
-        this.$container.empty();
+        this.$users_matching_view_container.empty();
+        this.users_matching_view_ids = [];
+        this.$other_users_container.empty();
+        this.other_user_ids = [];
+
+        // Reset data to be relevant for this current view.
+        this.set_render_data();
 
         // We rely on our caller to give us items
         // in already-sorted order.
         this.all_user_ids = opts.all_user_ids;
 
         this.fill_screen_with_content();
+        if (!this.tooltips_initialized) {
+            this.initialize_tooltips();
+        }
+
+        $("#buddy-list-users-matching-view-container .view-all-subscribers-link").remove();
+        $("#buddy-list-other-users-container .view-all-users-link").remove();
+        if (!buddy_data.get_is_searching_users()) {
+            this.render_view_user_list_links();
+        }
+        this.render_section_headers();
+        if (!this.render_data.hide_headers) {
+            this.update_empty_list_placeholders();
+        }
+    }
+
+    update_empty_list_placeholders() {
+        const {has_inactive_users_matching_view, has_inactive_other_users} = this.render_data;
+
+        let empty_list_message;
+        const is_searching_users = buddy_data.get_is_searching_users();
+        if (is_searching_users) {
+            empty_list_message = $t({defaultMessage: "No matching users."});
+        } else if (has_inactive_other_users) {
+            empty_list_message = $t({defaultMessage: "No active users."});
+        } else {
+            empty_list_message = $t({defaultMessage: "None."});
+        }
+
+        $("#buddy-list-other-users").data("search-results-empty", empty_list_message);
+        if ($("#buddy-list-other-users .empty-list-message").length) {
+            const empty_list_widget = render_empty_list_widget_for_list({empty_list_message});
+            $("#buddy-list-other-users").empty();
+            $("#buddy-list-other-users").append(empty_list_widget);
+        }
+
+        if (!is_searching_users) {
+            if (has_inactive_users_matching_view) {
+                empty_list_message = $t({defaultMessage: "No active users."});
+            } else {
+                empty_list_message = $t({defaultMessage: "None."});
+            }
+        }
+
+        $("#buddy-list-users-matching-view").data("search-results-empty", empty_list_message);
+        if ($("#buddy-list-users-matching-view .empty-list-message").length) {
+            const empty_list_widget = render_empty_list_widget_for_list({empty_list_message});
+            $("#buddy-list-users-matching-view").empty();
+            $("#buddy-list-users-matching-view").append(empty_list_widget);
+        }
+    }
```

Comment:
This function is very hard to read. I think you should just have a single block of conditionals at the top that sets two variables, `matching_view_empty_list_message` and `other_users_empty_list_message` and then does something with those. As it is, I think you're doing some sort of interesting fallthrough approach between the two blocks of conditionals that is just really hard to reason about -- and every time I read this I expect the `empty_list_message` calculated at the start to be used for the matching/active section, but then you're doing it in the other order.

Reply:
Ah yeah I see what you mean. I've updated the `buddy list: Show 'No active users' for empty lists with inactive users.` commit to calculate the two empty list messages first, and then start with setting matching users (instead of other users).

Reply:
If we want to keep this commit separate, I probably should also swap some ordering in an earlier commit, but I also think it would probably make sense to just squash this into the main buddy list commit.

Reply:
Yeah just squash it.